index.js

```diff
@@ -0,0 +1,90 @@
+'use strict'
+
+const fp = require('fastify-plugin')
+const CSRF = require('csrf')
+const { Forbidden } = require('http-errors')
+
+async function csrfPlugin (fastify, opts) {
+  const tokens = new CSRF()
+
+  const cookieKey = opts.cookieKey || '_csrf'
+  const cookieOpts = opts.cookieOpts || { path: '/', sameSite: true }
+  const sessionKey = opts.sessionKey || '_csrf'
+  const getToken = opts.getToken || getTokenDefault
+  const sessionPlugin = opts.sessionPlugin || null
+
+  if (sessionPlugin === 'fastify-secure-session') {
+    fastify.decorateReply('generateCsrf', generateCsrfSecureSession)
+  } else if (sessionPlugin === 'fastify-session') {
+    fastify.decorateReply('generateCsrf', generateCsrfSession)
+  } else {
+    fastify.decorateReply('generateCsrf', generateCsrfCookie)
+  }
+
+  fastify.decorate('csrfProtection', csrfProtection)
+
+  async function generateCsrfCookie (opts) {
+    let secret = this.request.cookies[cookieKey]
+    if (!secret) {
+      secret = await tokens.secret()
+      this.setCookie(cookieKey, secret, Object.assign({}, cookieOpts, opts))
+    }
+    return tokens.create(secret)
+  }
+
+  async function generateCsrfSecureSession () {
+    let secret = this.request.session.get(sessionKey)
+    if (!secret) {
+      secret = await tokens.secret()
+      this.request.session.set(sessionKey, secret)
+    }
+    return tokens.create(secret)
+  }
+
+  async function generateCsrfSession () {
+    let secret = this.request.session[sessionKey]
+    if (!secret) {
+      secret = await tokens.secret()
+      this.request.session[sessionKey] = secret
+    }
+    return tokens.create(secret)
+  }
+
+  function csrfProtection (req, reply, next) {
+    const secret = getSecret(req)
+    if (!secret) {
+      req.log.warn('Missing csrf secret')
+      return reply.send(new Forbidden('Missing csrf secret'))
+    }
+    if (!tokens.verify(secret, getToken(req))) {
+      req.log.warn('Invalid csrf token')
+      return reply.send(new Forbidden('Invalid csrf token'))
+    }
+    next()
+  }
+
+  function getSecret (req) {
+    if (sessionPlugin === 'fastify-secure-session') {
+      return req.session.get(sessionKey)
+    } else if (sessionPlugin === 'fastify-session') {
+      return req.session[sessionKey]
+    } else {
+      return req.cookies[cookieKey]
+    }
+  }
```

Comment:
Can you split this across different functions so the if is avoided at runtime? Can we automatically detect this instead of relying on a config value?

Reply:
> Can you split this across different functions so the if is avoided at runtime?

That was my initial implementation, but it's a lot of code duplication for little benefit.

> Can we automatically detect this instead of relying on a config value?

Probably yes, but it would be significantly slower than the current implementation.

Reply:
> Probably yes, but it would be significantly slower than the current implementation.

Not if you do it when the plugin is created.

---

Anyway, this is good to go.